### PR TITLE
STV Logic

### PR DESCRIPTION
**Pull request by SmithPeder**
This adds most of the backed logic _(except the STV algorithm)_ 

This builds on https://github.com/webkom/vote/pull/370, and must be merged after. The base branch for this commit is set to `STV-algorithm`

This changes the schema to look like this, where an election has `votes` and `seats`, and a vote has a list of priorities. The important change is that we no longer consider one Vote as a single vote for one person, but rather that a vote is a ballot containing a ranked list of alternatives.

image

#### Todos

- [x] There are more test/edge cases that should be implemented for the API changes
- [x] Fix failing race condition test
- [x] New tests for seats and strict

### CODE REVIEW

**Comment by SmithPeder** [comment]
I don't think we should implement the E2E tests in this PR, so those tests will keep failing. I think it's more natural that a `frontend/view` PR fixes the UI and the E2E protractor tests inn go.

File: test/api/vote.test.js [hunk 1]
```diff
@@ -201,20 +310,27 @@ describe('Vote API', () => {
...
-
     receivedVote.hash.should.equal(vote.hash);
-    receivedVote.alternative._id.should.equal(String(vote.alternative));
-    receivedVote.alternative.election.should.deep.equal({
-      _id: String(this.activeElection.id),
-      title: this.activeElection.title,
-    });
+  });
+
+  it('should be possible to retrieve a vote with correct election', async function () {
```
Comment (LudvigHz) [inline test/api/vote.test.js:323]: Can you add a test for the received `priotities` as well?
I.e that the posted priorites match the retrieved.

File: test/api/vote.test.js [hunk 2]
```diff
@@ -75,72 +78,178 @@ describe('Vote API', () => {
     passportStub.uninstall();
   });
 
-  it('should not be possible to vote with an invalid ObjectId as alternativeId', async () => {
+  it('should not be possible to vote without election', async () => {
     const { body: error } = await request(app)
       .post('/api/vote')
-      .send(votePayload('bad alternative'))
-      .expect(404)
+      .send({ priorities: [] })
```
Comment (LudvigHz) [inline test/api/vote.test.js:84]: Could you add something here to make sure we encapsulate only the missing election?

File: test/api/vote.test.js [hunk 3]
```diff
@@ -75,72 +78,178 @@ describe('Vote API', () => {
...
+      .expect('Content-Type', /json/);
     error.status.should.equal(400);
-    error.message.should.equal('Missing property alternativeId from payload.');
+    error.message.should.equal(
+      'One or more alternatives does not exist on election.'
+    );
   });
 
-  it('should be able to vote on alternative', async function () {
+  it('should be able to vote on active election with an empty priority list', async function () {
```
Comment (LudvigHz) [inline test/api/vote.test.js:188]: Could you assert the status code of the successful requests as well?

File: test/api/vote.test.js [hunk 4]
```diff
@@ -175,7 +284,7 @@ describe('Vote API', () => {
     await this.user.save();
     const { body: error } = await request(app)
       .post('/api/vote')
-      .send(votePayload(this.activeAlternative.id))
+      .send(votePayload(this.activeElection, []))
```
[the remark this answers is not in the document]
Comment (SmithPeder) [reply test/api/vote.test.js:287]: An empty array is allowed

File: test/api/vote.test.js [hunk 5]
```diff
@@ -190,7 +299,7 @@ describe('Vote API', () => {
   it('should not be able to vote on a deactivated election', async function () {
     const { body: error } = await request(app)
       .post('/api/vote')
-      .send(votePayload(this.inactiveAlternative.id))
+      .send(votePayload(this.inactiveElection, []))
```
[the remark this answers is not in the document]
Comment (SmithPeder) [reply test/api/vote.test.js:302]: An empty array is allowed

File: test/api/vote.test.js [hunk 6]
```diff
@@ -153,15 +262,15 @@ describe('Vote API', () => {
...
+    const votes = await Vote.find({ election: this.activeElection._id });
     votes.length.should.equal(1);
   });
 
   it('should not be possible to vote without logging in', async function () {
     passportStub.logout();
     const { body: error } = await request(app)
       .post('/api/vote')
-      .send(votePayload(this.activeAlternative.id))
+      .send(votePayload({}, []))
```
[the remark this answers is not in the document]
Comment (SmithPeder) [reply test/api/vote.test.js:273]: Kinda agree, will change

File: test/api/vote.test.js (hunk 2 above)
Comment (SmithPeder) [reply test/api/vote.test.js:84]: hmm, this don't make any sense 🤷🏻  🙃

File: app/controllers/vote.js [hunk 7]
```diff
@@ -1,26 +1,50 @@
...
+  return Election.findById(req.body.election._id)
+    .then(async (election) => {
+      // Election does not exist
+      if (!election) throw new errors.NotFoundError('election');
+
+      // Priorities cant be longer then alternatives
+      if (priorities.length > election.alternatives.length) {
+        throw new errors.InvalidPrioritiesLengthError(priorities, election);
+      }
+
```
[the remark this answers is not in the document]
Comment (SmithPeder) [reply app/controllers/vote.js:33]: Well, that's kinda the point of priorities well isn't it! 🗑️

File: test/api/vote.test.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (LudvigHz) [reply test/api/vote.test.js:287]: True, never mind 😅

**Comment by SmithPeder** [comment]
This must be merged with failing CI, as the `protractor` tests must be fixed in another branch.
